File: src/config-xml.ts
import { XMLParser, XMLBuilder } from 'fast-xml-parser';

export interface PipelineDef {
  type: 'inline' | 'scm';
  script?: string;
  scriptPath?: string;
  lightweight?: boolean;
}

export interface Trigger {
  type: 'cron' | 'scmPoll';
  spec: string;
}

export type ParameterType = 'string' | 'boolean' | 'text' | 'choice' | 'password';

export interface Parameter {
  type: ParameterType;
  name: string;
  default?: string;
  description?: string;
  choices?: string[];
}

export interface BuildRetention {
  numToKeep?: number;
  daysToKeep?: number;
}

export interface JobConfig {
  type: 'pipeline' | 'freestyle' | 'multibranch' | 'unknown';
  description?: string;
  disabled?: boolean;
  concurrentBuilds?: boolean;
  definition?: PipelineDef;
  triggers?: Trigger[];
  parameters?: Parameter[];
  buildRetention?: BuildRetention;
}

const PARSER_OPTS = {
  ignoreAttributes: false,
  attributeNamePrefix: '@_',
  cdataPropName: '__cdata',
  parseTagValue: false,
  trimValues: true,
  isArray: (name: string) => ALWAYS_ARRAY.has(name),
};

const BUILDER_OPTS = {
  ignoreAttributes: false,
  attributeNamePrefix: '@_',
  cdataPropName: '__cdata',
  format: true,
  indentBy: '  ',
  suppressEmptyNode: false,
  processEntities: true,
};

const ALWAYS_ARRAY = new Set([
  'hudson.plugins.git.BranchSpec',
  'hudson.model.StringParameterDefinition',
  'hudson.model.BooleanParameterDefinition',
  'hudson.model.TextParameterDefinition',
  'hudson.model.ChoiceParameterDefinition',
  'hudson.model.PasswordParameterDefinition',
  'hudson.triggers.TimerTrigger',
  'hudson.triggers.SCMTrigger',
]);

const PARAM_CLASS: Record<ParameterType, string> = {
  string:  'hudson.model.StringParameterDefinition',
  boolean: 'hudson.model.BooleanParameterDefinition',
  text:    'hudson.model.TextParameterDefinition',
  choice:  'hudson.model.ChoiceParameterDefinition',
  password:'hudson.model.PasswordParameterDefinition',
};

const CLASS_TO_PARAM_TYPE: Record<string, ParameterType> = Object.fromEntries(
  Object.entries(PARAM_CLASS).map(([k, v]) => [v, k as ParameterType])
);

function str(v: unknown): string | undefined {
  if (v === undefined || v === null) return undefined;
  if (typeof v === 'object' && '__cdata' in (v as object)) return String((v as Record<string, unknown>).__cdata);
  return String(v);
}

function bool(v: unknown): boolean | undefined {
  if (v === undefined || v === null) return undefined;
  return String(v).toLowerCase() === 'true';
}

function num(v: unknown): number | undefined {
  const n = Number(v);
  return isNaN(n) ? undefined : n;
}

function getRoot(parsed: Record<string, unknown>): [string, Record<string, unknown>] {
  for (const key of Object.keys(parsed)) {
    if (key !== '?xml' && key !== '!doctype') return [key, parsed[key] as Record<string, unknown>];
  }
  return ['unknown', {}];
}

function jobTypeFromTag(tag: string): JobConfig['type'] {
  if (tag === 'flow-definition') return 'pipeline';
  if (tag === 'project') return 'freestyle';
  if (tag.includes('MultiBranch') || tag.includes('multibranch')) return 'multibranch';
  return 'unknown';
}

function extractTriggers(raw: Record<string, unknown> | undefined): Trigger[] {
  if (!raw) return [];
  const out: Trigger[] = [];
  const timers = (raw['hudson.triggers.TimerTrigger'] as Array<Record<string, unknown>> | undefined) ?? [];
  for (const t of timers) {
    const spec = str(t.spec);
    if (spec) out.push({ type: 'cron', spec });
  }
  const scm = (raw['hudson.triggers.SCMTrigger'] as Array<Record<string, unknown>> | undefined) ?? [];
  for (const t of scm) {
    const spec = str(t.spec);
    if (spec) out.push({ type: 'scmPoll', spec });
  }
  return out;
}

function extractParameters(props: Record<string, unknown> | undefined): Parameter[] {
  if (!props) return [];
  const paramProp = props['hudson.model.ParametersDefinitionProperty'] as Record<string, unknown> | undefined;
  if (!paramProp) return [];
  const defs = paramProp.parameterDefinitions as Record<string, unknown> | undefined;
  if (!defs) return [];

  const out: Parameter[] = [];
  for (const [xmlTag, paramType] of Object.entries(CLASS_TO_PARAM_TYPE)) {
    const items = (defs[xmlTag] as Array<Record<string, unknown>> | undefined) ?? [];
    for (const p of items) {
      const param: Parameter = { type: paramType, name: str(p.name) ?? '' };
      const def = str(p.defaultValue ?? p.default);
      if (def !== undefined) param.default = def;
      const desc = str(p.description);
      if (desc) param.description = desc;
      if (paramType === 'choice') {
        const rawChoices = str(p.choices);
        if (rawChoices) param.choices = rawChoices.split('\n').map(s => s.trim()).filter(Boolean);
      }
      out.push(param);
    }
  }
  return out;
}

function extractRetention(discarder: Record<string, unknown> | undefined): BuildRetention | undefined {
  if (!discarder) return undefined;
  const strategy = discarder.strategy as Record<string, unknown> | undefined;
  if (!strategy) return undefined;
  const numToKeep = num(strategy.numToKeep);
  const daysToKeep = num(strategy.daysToKeep);
  const result: BuildRetention = {};
  if (numToKeep !== undefined && numToKeep >= 0) result.numToKeep = numToKeep;
  if (daysToKeep !== undefined && daysToKeep >= 0) result.daysToKeep = daysToKeep;
  return Object.keys(result).length > 0 ? result : undefined;
}

export function parseJobConfig(xml: string): JobConfig {
  const parser = new XMLParser(PARSER_OPTS);
  const parsed = parser.parse(xml) as Record<string, unknown>;
  const [tag, root] = getRoot(parsed);
  const type = jobTypeFromTag(tag);
  const cfg: JobConfig = { type };

  const desc = str(root.description);
  if (desc) cfg.description = desc;

  const disabled = bool(root.disabled);
  if (disabled !== undefined) cfg.disabled = disabled;

  const concurrent = bool(root.concurrentBuild);
  if (concurrent !== undefined) cfg.concurrentBuilds = concurrent;

  // Pipeline definition
  if (type === 'pipeline') {
    const defn = root.definition as Record<string, unknown> | undefined;
    if (defn) {
      const cls = str(defn['@_class']) ?? '';
      if (cls.includes('CpsFlowDefinition')) {
        cfg.definition = { type: 'inline', script: str(defn.script) ?? '' };
      } else if (cls.includes('CpsScmFlowDefinition')) {
        cfg.definition = {
          type: 'scm',
          scriptPath: str(defn.scriptPath) ?? 'Jenkinsfile',
          lightweight: bool(defn.lightweight) ?? true,
        };
      }
    }
  }

  const triggers = extractTriggers(root.triggers as Record<string, unknown> | undefined);
  if (triggers.length > 0) cfg.triggers = triggers;

  const parameters = extractParameters(root.properties as Record<string, unknown> | undefined);
  if (parameters.length > 0) cfg.parameters = parameters;

  const retention = extractRetention(root.buildDiscarder as Record<string, unknown> | undefined);
  if (retention) cfg.buildRetention = retention;

  return cfg;
}

function buildTriggerXml(triggers: Trigger[]): Record<string, unknown> {
  const out: Record<string, unknown> = {};
  const crons = triggers.filter(t => t.type === 'cron');
  const polls = triggers.filter(t => t.type === 'scmPoll');
  if (crons.length > 0) out['hudson.triggers.TimerTrigger'] = crons.map(t => ({ spec: t.spec }));
  if (polls.length > 0) out['hudson.triggers.SCMTrigger'] = polls.map(t => ({ spec: t.spec, ignorePostCommitHooks: false }));
  return out;
}

function buildParameterXml(parameters: Parameter[]): Record<string, unknown> {
  const defs: Record<string, unknown[]> = {};
  for (const p of parameters) {
    const tag = PARAM_CLASS[p.type];
    if (!tag) continue;
    const entry: Record<string, unknown> = { name: p.name, defaultValue: p.default ?? '', description: p.description ?? '' };
    if (p.type === 'choice') entry.choices = (p.choices ?? []).join('\n');
    if (!defs[tag]) defs[tag] = [];
    (defs[tag] as unknown[]).push(entry);
  }
  return {
    'hudson.model.ParametersDefinitionProperty': {
      parameterDefinitions: defs,
    },
  };
}

export function mergeJobConfig(xml: string, patch: Partial<JobConfig>): string {
  const parser = new XMLParser(PARSER_OPTS);
  const parsed = parser.parse(xml) as Record<string, unknown>;
  const [tag, root] = getRoot(parsed);

  if (patch.description !== undefined) root.description = patch.description;
  if (patch.disabled !== undefined) root.disabled = patch.disabled;
  if (patch.concurrentBuilds !== undefined) root.concurrentBuild = patch.concurrentBuilds;

  if (patch.definition !== undefined && jobTypeFromTag(tag) === 'pipeline') {
    const existing = (root.definition ?? {}) as Record<string, unknown>;
    if (patch.definition.type === 'inline') {
      existing['@_class'] = 'org.jenkinsci.plugins.workflow.csd.CpsFlowDefinition';
      existing.script = patch.definition.script ?? '';
      existing.sandbox = true;
      delete existing.scm;
      delete existing.scriptPath;
      delete existing.lightweight;
    } else if (patch.definition.type === 'scm') {
      existing['@_class'] = 'org.jenkinsci.plugins.workflow.csd.CpsScmFlowDefinition';
      if (patch.definition.scriptPath !== undefined) existing.scriptPath = patch.definition.scriptPath;
      if (patch.definition.lightweight !== undefined) existing.lightweight = patch.definition.lightweight;
      delete existing.script;
      delete existing.sandbox;
    } else {
      if (patch.definition.scriptPath !== undefined) existing.scriptPath = patch.definition.scriptPath;
      if (patch.definition.script !== undefined) existing.script = patch.definition.script;
    }
    root.definition = existing;
  }

  if (patch.triggers !== undefined) {
    root.triggers = buildTriggerXml(patch.triggers);
  }

  if (patch.parameters !== undefined) {
    const props = (root.properties ?? {}) as Record<string, unknown>;
    if (patch.parameters.length === 0) {
      delete props['hudson.model.ParametersDefinitionProperty'];
    } else {
      Object.assign(props, buildParameterXml(patch.parameters));
    }
    root.properties = props;
  }

  if (patch.buildRetention !== undefined) {
    const { numToKeep = -1, daysToKeep = -1 } = patch.buildRetention;
    root.buildDiscarder = {
      strategy: {
        '@_class': 'hudson.tasks.LogRotator',
        numToKeep,
        daysToKeep,
        artifactNumToKeep: -1,
        artifactDaysToKeep: -1,
      },
    };
  }

  parsed[tag] = root;
  const builder = new XMLBuilder(BUILDER_OPTS);
  return `<?xml version="1.0" encoding="UTF-8"?>\n${builder.build({ [tag]: root }) as string}`;
}

File: configxml.go
package main

import (
	"regexp"
	"strconv"
	"strings"

	"github.com/beevik/etree"
)

// xmlDeclRe matches a leading XML declaration. Jenkins emits version 1.1, which
// Go's encoding/xml rejects, so the declaration is stripped before parsing and
// re-emitted verbatim on write.
var xmlDeclRe = regexp.MustCompile(`(?s)^\s*<\?xml.*?\?>`)

const defaultXMLDecl = "<?xml version='1.1' encoding='UTF-8'?>"

// splitXMLDecl separates a leading XML declaration from the document body.
func splitXMLDecl(xml string) (decl, body string) {
	if m := xmlDeclRe.FindString(xml); m != "" {
		return strings.TrimSpace(m), xml[len(m):]
	}
	return "", xml
}

// newConfigDoc parses a Jenkins config.xml into an etree document, tolerating
// XML 1.1 declarations and preserving CDATA blocks (pipeline scripts). It also
// returns the original declaration so writers can re-emit it.
func newConfigDoc(xml string) (*etree.Document, string, error) {
	decl, body := splitXMLDecl(xml)
	doc := etree.NewDocument()
	doc.ReadSettings = etree.ReadSettings{Permissive: true, PreserveCData: true}
	if err := doc.ReadFromString(body); err != nil {
		return nil, decl, err
	}
	return doc, decl, nil
}

// JobConfig is the lossy, structured view of a Jenkins job's config.xml that the
// tools expose. Reads project a subset of the XML; writes are merged back into
// the original document so unknown elements are preserved.
type JobConfig struct {
	Type             string          `json:"type"`
	Description      *string         `json:"description,omitempty"`
	Disabled         *bool           `json:"disabled,omitempty"`
	ConcurrentBuilds *bool           `json:"concurrentBuilds,omitempty"`
	Definition       *PipelineDef    `json:"definition,omitempty"`
	Triggers         []Trigger       `json:"triggers,omitempty"`
	Parameters       []Parameter     `json:"parameters,omitempty"`
	BuildRetention   *BuildRetention `json:"buildRetention,omitempty"`
}

type PipelineDef struct {
	Type        string `json:"type"` // "inline" | "scm"
	Script      string `json:"script,omitempty"`
	ScriptPath  string `json:"scriptPath,omitempty"`
	Lightweight *bool  `json:"lightweight,omitempty"`
}

type Trigger struct {
	Type string `json:"type"` // "cron" | "scmPoll"
	Spec string `json:"spec"`
}

type Parameter struct {
	Type        string   `json:"type"` // string|boolean|text|choice|password
	Name        string   `json:"name"`
	Default     *string  `json:"default,omitempty"`
	Description string   `json:"description,omitempty"`
	Choices     []string `json:"choices,omitempty"`
}

type BuildRetention struct {
	NumToKeep  *int `json:"numToKeep,omitempty"`
	DaysToKeep *int `json:"daysToKeep,omitempty"`
}

const (
	tagTimerTrigger = "hudson.triggers.TimerTrigger"
	tagSCMTrigger   = "hudson.triggers.SCMTrigger"
	tagParamsProp   = "hudson.model.ParametersDefinitionProperty"
)

// paramClass maps the public parameter type onto its Jenkins XML element name.
var paramClass = map[string]string{
	"string":   "hudson.model.StringParameterDefinition",
	"boolean":  "hudson.model.BooleanParameterDefinition",
	"text":     "hudson.model.TextParameterDefinition",
	"choice":   "hudson.model.ChoiceParameterDefinition",
	"password": "hudson.model.PasswordParameterDefinition",
}

// classToParamType is the reverse of paramClass.
var classToParamType = func() map[string]string {
	m := make(map[string]string, len(paramClass))
	for k, v := range paramClass {
		m[v] = k
	}
	return m
}()

func jobTypeFromTag(tag string) string {
	switch {
	case tag == "flow-definition":
		return "pipeline"
	case tag == "project":
		return "freestyle"
	case strings.Contains(strings.ToLower(tag), "multibranch"):
		return "multibranch"
	default:
		return "unknown"
	}
}

// parseJobConfig reads a Jenkins config.xml into the structured JobConfig view.
func parseJobConfig(xml string) (*JobConfig, error) {
	doc, _, err := newConfigDoc(xml)
	if err != nil {
		return nil, err
	}
	root := doc.Root()
	if root == nil {
		return &JobConfig{Type: "unknown"}, nil
	}
	cfg := &JobConfig{Type: jobTypeFromTag(root.Tag)}

	if d := childText(root, "description"); d != "" {
		cfg.Description = &d
	}
	if b, ok := childBool(root, "disabled"); ok {
		cfg.Disabled = &b
	}
	if b, ok := childBool(root, "concurrentBuild"); ok {
		cfg.ConcurrentBuilds = &b
	}

	if cfg.Type == "pipeline" {
		if defn := root.SelectElement("definition"); defn != nil {
			class := defn.SelectAttrValue("class", "")
			switch {
			case strings.Contains(class, "CpsFlowDefinition"):
				cfg.Definition = &PipelineDef{Type: "inline", Script: childText(defn, "script")}
			case strings.Contains(class, "CpsScmFlowDefinition"):
				lightweight := true
				if b, ok := childBool(defn, "lightweight"); ok {
					lightweight = b
				}
				scriptPath := childText(defn, "scriptPath")
				if scriptPath == "" {
					scriptPath = "Jenkinsfile"
				}
				cfg.Definition = &PipelineDef{Type: "scm", ScriptPath: scriptPath, Lightweight: &lightweight}
			}
		}
	}

	if trig := root.SelectElement("triggers"); trig != nil {
		for _, t := range trig.SelectElements(tagTimerTrigger) {
			if spec := childText(t, "spec"); spec != "" {
				cfg.Triggers = append(cfg.Triggers, Trigger{Type: "cron", Spec: spec})
			}
		}
		for _, t := range trig.SelectElements(tagSCMTrigger) {
			if spec := childText(t, "spec"); spec != "" {
				cfg.Triggers = append(cfg.Triggers, Trigger{Type: "scmPoll", Spec: spec})
			}
		}
	}

	cfg.Parameters = parseParameters(root)

	if bd := root.SelectElement("buildDiscarder"); bd != nil {
		if strat := bd.SelectElement("strategy"); strat != nil {
			ret := BuildRetention{}
			if n, ok := childInt(strat, "numToKeep"); ok && n >= 0 {
				ret.NumToKeep = &n
			}
			if n, ok := childInt(strat, "daysToKeep"); ok && n >= 0 {
				ret.DaysToKeep = &n
			}
			if ret.NumToKeep != nil || ret.DaysToKeep != nil {
				cfg.BuildRetention = &ret
			}
		}
	}

	return cfg, nil
}

func parseParameters(root *etree.Element) []Parameter {
	props := root.SelectElement("properties")
	if props == nil {
		return nil
	}
	paramProp := props.SelectElement(tagParamsProp)
	if paramProp == nil {
		return nil
	}
	defs := paramProp.SelectElement("parameterDefinitions")
	if defs == nil {
		return nil
	}
	var out []Parameter
	for class, ptype := range classToParamType {
		for _, p := range defs.SelectElements(class) {
			param := Parameter{Type: ptype, Name: childText(p, "name")}
			if d := p.SelectElement("defaultValue"); d != nil {
				v := d.Text()
				param.Default = &v
			} else if d := p.SelectElement("default"); d != nil {
				v := d.Text()
				param.Default = &v
			}
			if desc := childText(p, "description"); desc != "" {
				param.Description = desc
			}
			if ptype == "choice" {
				param.Choices = parseChoices(p)
			}
			out = append(out, param)
		}
	}
	return out
}

// parseChoices extracts choice strings from a ChoiceParameterDefinition. Jenkins
// stores them as <choices><a><string>x</string>...</a></choices>; older configs
// may store a newline-separated text body. Both are handled.
func parseChoices(p *etree.Element) []string {
	choices := p.SelectElement("choices")
	if choices == nil {
		return nil
	}
	var out []string
	if a := choices.SelectElement("a"); a != nil {
		for _, s := range a.SelectElements("string") {
			if v := strings.TrimSpace(s.Text()); v != "" {
				out = append(out, v)
			}
		}
	}
	if len(out) == 0 {
		for _, line := range strings.Split(choices.Text(), "\n") {
			if v := strings.TrimSpace(line); v != "" {
				out = append(out, v)
			}
		}
	}
	return out
}

// mergeJobConfig applies patch onto the original config.xml, preserving every
// element it does not touch, and returns the serialized result.
func mergeJobConfig(xml string, patch *JobConfig) (string, error) {
	doc, decl, err := newConfigDoc(xml)
	if err != nil {
		return "", err
	}
	root := doc.Root()
	if root == nil {
		return "", errEmptyConfig
	}

	if patch.Description != nil {
		setChildText(root, "description", *patch.Description)
	}
	if patch.Disabled != nil {
		setChildText(root, "disabled", strconv.FormatBool(*patch.Disabled))
	}
	if patch.ConcurrentBuilds != nil {
		setChildText(root, "concurrentBuild", strconv.FormatBool(*patch.ConcurrentBuilds))
	}

	if patch.Definition != nil && jobTypeFromTag(root.Tag) == "pipeline" {
		mergeDefinition(root, patch.Definition)
	}

	if patch.Triggers != nil {
		removeChild(root, "triggers")
		trig := root.CreateElement("triggers")
		for _, t := range patch.Triggers {
			switch t.Type {
			case "cron":
				el := trig.CreateElement(tagTimerTrigger)
				el.CreateElement("spec").SetText(t.Spec)
			case "scmPoll":
				el := trig.CreateElement(tagSCMTrigger)
				el.CreateElement("spec").SetText(t.Spec)
				el.CreateElement("ignorePostCommitHooks").SetText("false")
			}
		}
	}

	if patch.Parameters != nil {
		props := root.SelectElement("properties")
		if props == nil {
			props = root.CreateElement("properties")
		}
		removeChild(props, tagParamsProp)
		if len(patch.Parameters) > 0 {
			prop := props.CreateElement(tagParamsProp)
			defs := prop.CreateElement("parameterDefinitions")
			for _, p := range patch.Parameters {
				class := paramClass[p.Type]
				if class == "" {
					continue
				}
				el := defs.CreateElement(class)
				el.CreateElement("name").SetText(p.Name)
				def := ""
				if p.Default != nil {
					def = *p.Default
				}
				el.CreateElement("defaultValue").SetText(def)
				el.CreateElement("description").SetText(p.Description)
				if p.Type == "choice" {
					choices := el.CreateElement("choices")
					choices.CreateAttr("class", "java.util.Arrays$ArrayList")
					a := choices.CreateElement("a")
					a.CreateAttr("class", "string-array")
					for _, c := range p.Choices {
						a.CreateElement("string").SetText(c)
					}
				}
			}
		}
	}

	if patch.BuildRetention != nil {
		removeChild(root, "buildDiscarder")
		bd := root.CreateElement("buildDiscarder")
		bd.CreateAttr("class", "hudson.tasks.LogRotator")
		strat := bd.CreateElement("strategy")
		strat.CreateAttr("class", "hudson.tasks.LogRotator")
		num, days := -1, -1
		if patch.BuildRetention.NumToKeep != nil {
			num = *patch.BuildRetention.NumToKeep
		}
		if patch.BuildRetention.DaysToKeep != nil {
			days = *patch.BuildRetention.DaysToKeep
		}
		strat.CreateElement("daysToKeep").SetText(strconv.Itoa(days))
		strat.CreateElement("numToKeep").SetText(strconv.Itoa(num))
		strat.CreateElement("artifactDaysToKeep").SetText("-1")
		strat.CreateElement("artifactNumToKeep").SetText("-1")
	}

	doc.Indent(2)
	out, err := doc.WriteToString()
	if err != nil {
		return "", err
	}
	if decl == "" {
		decl = defaultXMLDecl
	}
	return decl + "\n" + strings.TrimLeft(out, "\n"), nil
}

func mergeDefinition(root *etree.Element, def *PipelineDef) {
	existing := root.SelectElement("definition")
	if existing == nil {
		existing = root.CreateElement("definition")
	}
	switch def.Type {
	case "inline":
		setAttr(existing, "class", "org.jenkinsci.plugins.workflow.cps.CpsFlowDefinition")
		setChildText(existing, "script", def.Script)
		setChildText(existing, "sandbox", "true")
		removeChild(existing, "scm")
		removeChild(existing, "scriptPath")
		removeChild(existing, "lightweight")
	case "scm":
		setAttr(existing, "class", "org.jenkinsci.plugins.workflow.cps.CpsScmFlowDefinition")
		if def.ScriptPath != "" {
			setChildText(existing, "scriptPath", def.ScriptPath)
		}
		if def.Lightweight != nil {
			setChildText(existing, "lightweight", strconv.FormatBool(*def.Lightweight))
		}
		removeChild(existing, "script")
		removeChild(existing, "sandbox")
	default:
		if def.ScriptPath != "" {
			setChildText(existing, "scriptPath", def.ScriptPath)
		}
		if def.Script != "" {
			setChildText(existing, "script", def.Script)
		}
	}
}

// ── etree helpers ───────────────────────────────────────────────────────────

func childText(e *etree.Element, tag string) string {
	if c := e.SelectElement(tag); c != nil {
		return c.Text()
	}
	return ""
}

func childBool(e *etree.Element, tag string) (bool, bool) {
	c := e.SelectElement(tag)
	if c == nil {
		return false, false
	}
	return strings.EqualFold(strings.TrimSpace(c.Text()), "true"), true
}

func childInt(e *etree.Element, tag string) (int, bool) {
	c := e.SelectElement(tag)
	if c == nil {
		return 0, false
	}
	n, err := strconv.Atoi(strings.TrimSpace(c.Text()))
	if err != nil {
		return 0, false
	}
	return n, true
}

func setChildText(e *etree.Element, tag, value string) {
	c := e.SelectElement(tag)
	if c == nil {
		c = e.CreateElement(tag)
	}
	c.SetText(value)
}

func setAttr(e *etree.Element, key, value string) {
	if a := e.SelectAttr(key); a != nil {
		a.Value = value
		return
	}
	e.CreateAttr(key, value)
}

func removeChild(e *etree.Element, tag string) {
	for _, c := range e.SelectElements(tag) {
		e.RemoveChild(c)
	}
}
